### DISCUSSION TITLE
Get-LcsDatabaseOperationStatus - "Object reference not set to an instance of an object"

### DISCUSSION THREAD

**Discussion by assofohdz** [d365collaborative/d365fo.tools#529] [Q&A] [answered]
So, in stage one, I have a db refresh request sent with:

`$databaseRefresh, $status = Invoke-D365LcsDatabaseRefresh -SourceEnvironmentId "$(LCS-EnvId-PROD)" -TargetEnvironmentId "$(LCS-EnvId-PT05)" -Verbose`

After that, I persist `$status.OperationActivityId` into a release variable `dbRefreshActivityId` and would then like to (in the next stage) poll the status using this operation activity id. That, I do like this:

```
do {
	Get-D365LcsApiConfig | Invoke-D365LcsApiRefreshToken -ClientId "$(LCS-client-id)" | Set-D365LcsApiConfig

        $databaseOperationStatus = Get-D365LcsDatabaseOperationStatus -EnvironmentId "$(LCS-EnvId-PT05)" -OperationActivityId "$(dbRefreshActivityId)" -EnableException -Verbose

	Write-PSFMessage -Level Verbose -Message "Database Operation Status is: $($databaseOperationStatus.OperationStatus)"
}
while ((($databaseOperationStatus.OperationStatus -eq "InProgress") -or ($databaseOperationStatus.OperationStatus -eq "NotStarted") -or ($databaseOperationStatus.OperationStatus -eq "RollbackInProgress")))
```
This should refresh token, and get the status after the default sleep time of 300 seconds, but, it comes back saying (dont mind the 5 second sleep, that's just for testing)

> 2021-03-29T11:39:01.8806556Z **ActivityId: a6918ac1-f137-4a0b-ab9f-05c327d8b1ff**
> 2021-03-29T11:39:02.1241597Z VERBOSE: [11:39:02][Get-D365LcsDatabaseOperationStatus] Sleeping before hitting the LCS API for Deployment Status
> 2021-03-29T11:39:07.1555986Z VERBOSE: [11:39:07][Get-LcsDatabaseOperationStatus] Json payload for LCS generated.
> 2021-03-29T11:39:07.1721694Z VERBOSE: [11:39:07][New-JsonRequest] Building a HttpRequestMessage.
> 2021-03-29T11:39:07.1777939Z VERBOSE: [11:39:07][New-JsonRequest] Adding Authorization token to the HttpRequestMessage.
> 2021-03-29T11:39:07.1916262Z VERBOSE: [11:39:07][Get-LcsDatabaseOperationStatus] Invoke LCS request.
> 2021-03-29T11:39:07.2068130Z VERBOSE: [11:39:07][Get-AsyncResult] Building the Task Waiter and start waiting.
> 2021-03-29T11:39:08.2945140Z VERBOSE: [11:39:08][Get-LcsDatabaseOperationStatus] Extracting the response received from LCS.
> 2021-03-29T11:39:08.3049578Z VERBOSE: [11:39:08][Get-AsyncResult] Building the Task Waiter and start waiting.
> 2021-03-29T11:39:08.3233431Z VERBOSE: [11:39:08][Get-LcsDatabaseOperationStatus] Extracting the response received from LCS.
> 2021-03-29T11:39:08.3425275Z [11:39:08][Get-LcsDatabaseOperationStatus] Unknown error requesting database refresh status.
> 2021-03-29T11:39:08.3536013Z [11:39:08][Get-LcsDatabaseOperationStatus] Error in request for database refresh status of environment: 'Object reference not set to an instance of an object.' **(Activity Id: 'acad9d32-739a-487e-907c-3ef43858f93b')**
> 

The first highlighted activity, is my persisted variable. The second activity is a new one each time - what am I doing wrong here?

**Comment by Splaxi** [comment]
Did you get the script to work when running locally on a machine first?

Can you print the content of the release variable, dbRefreshActivityId, in the next stage? 

You could try to output the `$error` variable - it might contain details that will show you what is missing.

**Reply by assofohdz** [reply]
Yep, the first highlighted activityid is the variable that was persisted in the first stage, so that's working

I'll try the $error variable

**Reply by Splaxi** [reply]
I would personally wait 60 seconds, from the initial invoke of the database refresh, before doing a GetDatabaseStatus. We have seen issues with the LCS API if hitting it to fast / hard with requests.

I run Invoke-D365LcsDatabaseRefresh and Get-D365LcsDatabaseOperationStatus in the same stage, but other than that our solutions looks the same. And I pipe the result object from `Invoke-D365LcsDatabaseRefresh` directly to the `Get-D365LcsDatabaseOperationStatus` cmdlet.

But please note the `-SkipInitialStatusFetch` parameter for the `Invoke-D365LcsDatabaseRefresh` cmdlet. If you don't set that, the output will contain 2 results, which in powershell can lead to some troubles, because of how it handles arrays.

**Comment by assofohdz** [comment, marked as answer]
With help from @Splaxi I figured it out. I didnt know that the ActivityId and OperationActivityId of the Get-D365LcsDatabaseOperationStatus return object were not the same. So, once I started using the right (OperationActivityId) I got it working.
